test/Npgsql.Tests/ConnectionTests.cs

```diff
@@ -1395,6 +1395,49 @@ public async Task NoResetOnClose(bool noResetOnClose)
                 : originalApplicationName));
     }
 
+    [Test]
+    [Description("Test whether the internal NpgsqlConnection.Open method stays on the same thread with async=false")]
+    public async Task Sync_open_blocked_same_thread()
+    {
+        if (IsMultiplexing)
+            return;
+
+        await using var dataSource = CreateDataSource(csb =>
+        {
+            csb.MaxPoolSize = 1;
+        });
+
+        await using var openConnection = await dataSource.OpenConnectionAsync();
+
+        // 2 tasks are usually enough to reproduce the issue
```

Comment:
This sounds flaky?

Shouldn't it be possible to assert with one task? Given we're already at pool max?

Instead of immediately awaiting the conn.Open you should be able to add a thread.sleep after it so task stealing for the taskscheduler item kicks in. That should be a fairly reliable way to get it onto another thread?

Reply:
To be fair in all my runs it has always failed (unlike if there was only a single task), so `usually enough` is a bit incorrect here.

> Instead of immediately awaiting the conn.Open you should be able to add a thread.sleep after it so task stealing for the taskscheduler item kicks in (and only then await). That should be a fairly reliable way to get it onto another thread?

OK, will try that later today.

Reply:
Nope, still can't repro with `Thread.Sleep` added. We might as well just assert that task from `conn.Open` has run to completion before awaiting it.

Reply:
One thing we might do is to replace `GetAwaiter.GetResult` calls in our codebase with something like:

```csharp
internal static T CompleteSynchronously<T>(this Task<T> task)
{
	Debug.Assert(task.Status == TaskStatus.RanToCompletion);
	return task.GetAwaiter().GetResult();
}
```